internal/internal_worker_cache.go

```diff
@@ -0,0 +1,124 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"runtime"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"go.temporal.io/sdk/internal/common/cache"
+)
+
+// A shared cache workers can use to store workflow state. The cache is expected to be initialized with the first worker
+// to be instantiated, and freed with the last worker to terminate. IE: All workers have a pointer to it.
+type workerCache struct {
+	workflowCache *cache.Cache
+}
+
+// Don't touch this except via methods on workerCache. Holds shared workflow cache.
+var workflowCache cache.Cache
+var stickyCacheSize = defaultStickyCacheSize
+var stickyCacheLock sync.Mutex
+
+// Holds the number of outstanding references to the cache so we can zero it out if all workers are dead
+var workerRefcount atomic.Int32
+
+// SetStickyWorkflowCacheSize sets the cache size for sticky workflow cache. Sticky workflow execution is the affinity
+// between workflow tasks of a specific workflow execution to a specific worker. The affinity is set if sticky execution
+// is enabled via Worker.Options (It is enabled by default unless disabled explicitly). The benefit of sticky execution
+// is that workflow does not have to reconstruct the state by replaying from beginning of history events. But the cost
+// is it consumes more memory as it rely on caching workflow execution's running state on the worker. The cache is shared
+// between workers running within same process. This must be called before any worker is started. If not called, the
+// default size of 10K (might change in future) will be used.
+func SetStickyWorkflowCacheSize(cacheSize int) {
+	stickyCacheLock.Lock()
+	defer stickyCacheLock.Unlock()
+	if workflowCache != nil {
+		panic("cache already created, please set cache size before worker starts.")
+	}
+	stickyCacheSize = cacheSize
+}
+
+// Returns a pointer to the workerCache, and increases the outstanding refcount by one. Instances of
+// workerCache decrement the refcounter as a hook to runtime.SetFinalizer
```

Comment:
I would rename it to something that indicates that it is not just get, but actually more like lease that should be returned.

Reply:
I changed the semantics just a little to something that makes a bit more sense. Let me know what you think now.